### PR TITLE
Nft/service

### PR DESCRIPTION
**Pull request by FrozenKiwi**
Implement a simple online service to handle IPFS uploads and gass-less updates of our NFT's.

### CODE REVIEW

File: apps/nft-service/src/init/index.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { init as initlog, log } from '@thecointech/logging';
+
+export async function init() {
+  // Logging (todo: connect to logs aggregator)
+  initlog('nft-service');
+  log.trace('Initialization complete')
```
Comment (FrozenKiwi) [inline apps/nft-service/src/init/index.ts:6]: We probably should see an account on dev:live with an NFT

File: apps/nft-service/src/main.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import swaggerUi from 'swagger-ui-express';
+import swaggerDocument from './api/swagger.json';
+import { init } from './init';
+import cors from 'cors';
+;
+
+const app = express();
+// enable cors
+app.use(cors());
+const port = process.env.PORT ?? 9009;
```
Comment (FrozenKiwi) [inline apps/nft-service/src/main.ts:12]: TODO: Normalize ports etc

**Review by MarieTheCoin — APPROVED** [review]

File: apps/nft-service/src/init/index.ts (hunk 1 above)
Comment (FrozenKiwi) [reply apps/nft-service/src/init/index.ts:6]: Seed :-).  Misspelled - seeding is added in a later PR

File: apps/nft-service/src/main.ts (hunk 2 above)
Comment (FrozenKiwi) [reply apps/nft-service/src/main.ts:12]: Done in subsequent PR

File: apps/nft-service/src/nft/gassless.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    log.warn({TokenID: request.tokenId}, `Invalid request to update {TokenID}: signature ${address} does not match owner`);
+    return true;
+  }
+  return false;
+}
+
+const isTimeLocked = async (contract: TheCoinNFT, tokenId: number) => {
+  const canUpdate = contract.canUpdate(tokenId);
+  if (!canUpdate) {
+    log.warn({TokenID: tokenId}, `Invalid request to update {TokenID}: NFT is still timelocked`);
```
[the remark this answers is not in the document]
Comment (FrozenKiwi) [reply apps/nft-service/src/nft/gassless.ts:37]: Uggghhh, I hate writing error messages

Yes, that's it.  Timelocked sounded good at the time, but now it just doesn't seem right.  Recharged?  Maybe rewrite to say: "NFT can be updated again after ${date}" ?

File: apps/nft-service/src/nft/validate.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+// Basic validation of the JSON data present
+async function isValidJson(json: MetadataJson, address: string) {
+  // Is this too big?
+  if (JSON.stringify(json).length > MAX_JSON_LENGTH) {
+    log.warn({ address }, `Rejected upload from {address} because an JSON payload is too big`);
+    return false;
+  }
+  // TODO: This validates CIDv0 - We probably want to switch to CIDv1 for browser interop
+  if (!json.hash.match(/^\/ipfs\/[123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz]{46}$/)) {
+    log.warn({ address }, `Rejected upload from {address} because an JSON payload is too big`);
```
[the remark this answers is not in the document]
Comment (FrozenKiwi) [reply apps/nft-service/src/nft/validate.ts:67]: nope :-)

[the remark this answers is not in the document]
Comment (FrozenKiwi) [reply apps/nft-service/src/nft/validate.ts:67]: Fixed, thanks

**Comment by FrozenKiwi** [comment]
Ignoring DC error (it is fixed in a later PR) and merging

**Pull request merged**